src/ansi-c/c_typecheck_expr.cpp

```diff
@@ -3468,3 +3445,355 @@ void c_typecheck_baset::make_constant_index(exprt &expr)
     throw 0;
   }
 }
+
+optionalt<symbol_exprt> c_typecheck_baset::typecheck_gcc_polymorphic_builtin(
+  const irep_idt &identifier,
+  const exprt::operandst &arguments,
+  const source_locationt &source_location)
+{
+  // the arguments need not be type checked just yet, thus do not make
+  // assumptions that would require type checking
+
+  if(
+    identifier == "__sync_fetch_and_add" ||
+    identifier == "__sync_fetch_and_sub" ||
+    identifier == "__sync_fetch_and_or" ||
+    identifier == "__sync_fetch_and_and" ||
+    identifier == "__sync_fetch_and_xor" ||
+    identifier == "__sync_fetch_and_nand" ||
+    identifier == "__sync_add_and_fetch" ||
+    identifier == "__sync_sub_and_fetch" ||
+    identifier == "__sync_or_and_fetch" ||
+    identifier == "__sync_and_and_fetch" ||
+    identifier == "__sync_xor_and_fetch" ||
+    identifier == "__sync_nand_and_fetch" ||
+    identifier == "__sync_bool_compare_and_swap" ||
+    identifier == "__sync_val_compare_and_swap" ||
+    identifier == "__sync_lock_test_and_set" ||
+    identifier == "__sync_lock_release")
+  {
+    // These are polymorphic, see
+    // http://gcc.gnu.org/onlinedocs/gcc-4.1.1/gcc/Atomic-Builtins.html
+
+    // adjust return type of function to match pointer subtype
+    if(arguments.empty())
+    {
+      error().source_location = source_location;
+      error() << "__sync_* primitives take as least one argument" << eom;
+      throw 0;
+    }
+
+    const exprt &ptr_arg = arguments.front();
+
+    if(ptr_arg.type().id() != ID_pointer)
+    {
+      error().source_location = source_location;
+      error() << "__sync_* primitives take a pointer as first argument" << eom;
+      throw 0;
+    }
+
+    typet return_type = ptr_arg.type().subtype();
+    if(identifier == "__sync_bool_compare_and_swap")
+      return_type = c_bool_type();
+    else if(identifier == "__sync_lock_release")
+      return_type = void_type();
+
+    code_typet t{{code_typet::parametert(ptr_arg.type())}, return_type};
+    t.make_ellipsis();
+    symbol_exprt result{identifier, t};
+    result.add_source_location() = source_location;
```

Comment:
:see_no_evil:

Reply:
Are you suggesting that `symbol_exprt` should also have a constructor that takes a `source_locationt`?

Reply:
Not necessarily, but that is a side discussion. Nothing for this PR.
I'd expect either `result.add_source_location(source_location)` (If it is something to be "added") or
probably `result.set_source_location(source_location)` or
maybe `result.source_location() = source_location`.